lib/nerve/reporter/zookeeper.rb

```diff
@@ -145,33 +159,113 @@ def encode_child_name(service)
-          @zk.delete(@full_key, :ignore => :no_node)
+          @zk.delete(node_path, :ignore => :no_node)
         end
-        @full_key = nil
+
+        @full_key.set(nil)
       end
     end
 
     def zk_create
       # only mkdir_p if the path does not exist
       statsd.time('nerve.reporter.zk.create.elapsed_time', tags: ["zk_cluster:#{@zk_cluster}", "zk_path:#{@zk_path}"]) do
         @zk.mkdir_p(@zk_path) unless @zk.exists?(@zk_path)
-        @full_key = @zk.create(@key_prefix, :data => @data, :mode => @mode)
+
+        node_path = zk_try_create
+        @full_key.set(node_path)
+        log.info "nerve: wrote new ZK node of type #{@mode} at #{node_path}"
       end
     end
 
-    def zk_save
-      return zk_create unless @full_key
+    def zk_try_create
+      begin
+        return @zk.create(@key_prefix, :data => @data, :mode => @mode)
+      rescue ::Zookeeper::Exceptions::NodeExists, ZK::Exceptions::NodeExists
+        # This exception will only occur when not using sequential
+        # nodes (because sequential nodes are always unique), in which
+        # case the name is the same as @key_prefix as Zookeeper
+        # will not append any suffix.
+        @zk.set(@key_prefix, @data)
+        log.info "nerve: tried to write node but exists, setting data instead"
+
+        return @key_prefix
+      end
+    end
+
+    def zk_save(node_path)
+      return zk_create unless node_path
 
       begin
         statsd.time('nerve.reporter.zk.save.elapsed_time', tags: ["zk_cluster:#{@zk_cluster}"]) do
-          @zk.set(@full_key, @data)
+          @zk.set(node_path, @data)
+          log.info "nerve: set data on #{node_path}"
         end
       rescue ZK::Exceptions::NoNode
         zk_create
       end
     end
+
+    def start_ttl_renew_thread
+      @ttl_should_exit = Nerve::AtomicValue.new(false)
+      @ttl_thread = nil
+
+      unless @node_ttl.nil? || @mode.to_s.start_with?('ephemeral')
+        @ttl_thread = Thread.new {
+          log.info "nerve: ttl renew: background thread starting"
+          last_run = Time.now - rand(@node_ttl)
+
+          until @ttl_should_exit.get
+            last_run = renew_ttl(last_run, Time.now)
+            sleep 0.5
+          end
+
+          log.info "synapse: ttl renew: background thread exiting normally"
+        }
+      end
+    end
+
+    # Renew the TTL of @full_key if more than @node_ttl seconds has passed
+    # between `now` and `last_refresh`.
+    # Returns the last refresh time *after performing the renewal.*
+    # If the TTL *is* renewed, it will return `now`.
+    # Otherwise, it will return `last_refresh`.
+    def renew_ttl(last_refresh, now)
+      elapsed = now - last_refresh
+
+      if elapsed >= @node_ttl
+        node_path = @full_key.get
+
+        if node_path.nil?
+          log.info "nerve: ttl renew: not touching ZK node because path not set"
+        else
+          begin
+            @zk.set(node_path, @data)
```

Comment:
assume this operation is fast enough? otherwise, the real renewed time could have a large delta with `now`

Reply:
Normally it is very fast. During heavy Zookeeper load, it is possible this can timeout (I believe?). I can send the latest time instead of the provided time